File: database-model/models/sport-club-court/court.py
from sqlmodel import SQLModel, Field
from sqlalchemy import CheckConstraint

class Court(SQLModel, table=True):
    __tablename__ = "court" # type: ignore
    # A court belongs EITHER to a club (public, selectable by anyone) OR to a
    # user (private, only selectable by its owner). Exactly one FK must be set.
    __table_args__ = (
        CheckConstraint(
            "(club_id IS NOT NULL AND owner_id IS NULL) "
            "OR (club_id IS NULL AND owner_id IS NOT NULL)",
            name="court_club_xor_owner",
        ),
    )
    id: int = Field(primary_key=True, index=True)
    club_id: int | None = Field(foreign_key="club.id", default=None)
    owner_id: int | None = Field(foreign_key="auth_user.id", default=None)
    sport_id: int = Field(foreign_key="sports.id")
    name: str = Field(max_length=50)
    price_per_hour: float
    is_indoor: bool = False
    lat: float | None = Field(default=None)
    lon: float | None = Field(default=None)

File: gohan/lib/features/courts/data/court.dart
/// Matches the backend `CourtOutputDTO` returned by `/api/courts/`.
///
/// A court is either a club court (club_id set, owner_id null) or a private
/// one (owner_id set, club_id null). The XOR is enforced DB-side.
class Court {
  final int id;
  final int? clubId;
  final int? ownerId;
  final int sportId;
  final String name;
  final double pricePerHour;
  final bool isIndoor;
  final double? lat;
  final double? lon;

  const Court({
    required this.id,
    required this.clubId,
    required this.ownerId,
    required this.sportId,
    required this.name,
    required this.pricePerHour,
    required this.isIndoor,
    required this.lat,
    required this.lon,
  });

  factory Court.fromJson(Map<String, dynamic> json) => Court(
        id: json['id'] as int,
        clubId: json['club_id'] as int?,
        ownerId: json['owner_id'] as int?,
        sportId: json['sport_id'] as int,
        name: json['name'] as String,
        pricePerHour: (json['price_per_hour'] as num).toDouble(),
        isIndoor: json['is_indoor'] as bool,
        lat: (json['lat'] as num?)?.toDouble(),
        lon: (json['lon'] as num?)?.toDouble(),
      );

  bool get isPrivate => ownerId != null;
}
